mono/utils/mono-cgroup.c

```diff
@@ -0,0 +1,958 @@
+static gboolean getPhysicalMemoryUsage(size_t *);
+static int findCGroupVersion(void);
+static gboolean isCGroup1MemorySubsystem(const char *);
+static gboolean isCGroup1CpuSubsystem(const char *);
+static char *findCGroupPath(gboolean (*is_subsystem)(const char *));
+static void findHierarchyMount(gboolean (*is_subsystem)(const char *), char **, char **);
+static char *findCGroupPathForSubsystem(gboolean (*is_subsystem)(const char *));
+static gboolean getCGroupMemoryLimit(guint64 *, const char *);
+static gboolean getCGroupMemoryUsage(size_t *);
+static size_t getPhysicalMemoryTotal(guint64);
+static long long readCpuCGroupValue(const char *);
+static void computeCpuLimit(long long, long long, uint32_t *);
+
+size_t getRestrictedPhysicalMemoryLimit(void);
+gboolean getPhysicalMemoryUsed(size_t *);
+size_t getPhysicalMemoryAvail(void);
+gboolean getCpuLimit(guint *);
+static gboolean readLongLongValueFromFile(const char *, long long *);
+
+// the cgroup version number or 0 to indicate cgroups are not found or not enabled
+static int s_cgroup_version;
+
+static char *s_memory_cgroup_path = NULL;
+static char *s_cpu_cgroup_path = NULL;
+
+static const char *s_mem_stat_key_names[4];
+static size_t s_mem_stat_key_lengths[4];
+static size_t s_mem_stat_n_keys = 0;
+static long pageSize;
+
+/**
+ * @brief Initialize variables used by the calculation routines.
+ *
+ */
+static void 
+initialize()
+{
+	s_cgroup_version = findCGroupVersion();
+	s_memory_cgroup_path = findCGroupPath(s_cgroup_version == 1 ? &isCGroup1MemorySubsystem : NULL);
+	s_cpu_cgroup_path = findCGroupPath(s_cgroup_version == 1 ? &isCGroup1CpuSubsystem : NULL);
+
+	if (s_cgroup_version == 1) {
+		s_mem_stat_n_keys = 4;
+		s_mem_stat_key_names[0] = "total_inactive_anon ";
+		s_mem_stat_key_names[1] = "total_active_anon ";
+		s_mem_stat_key_names[2] = "total_dirty ";
+		s_mem_stat_key_names[3] = "total_unevictable ";
+	} else {
+		s_mem_stat_n_keys = 3;
+		s_mem_stat_key_names[0] = "anon ";
+		s_mem_stat_key_names[1] = "file_dirty ";
+		s_mem_stat_key_names[2] = "unevictable ";
+	}
+
+	for (size_t i = 0; i < s_mem_stat_n_keys; i++)
+		s_mem_stat_key_lengths[i] = strlen(s_mem_stat_key_names[i]);
+
+	pageSize = sysconf(_SC_PAGE_SIZE);
+}
+
+/**
+ *
+ * @brief Read a value from a specified /sys/fs/cgroup/memory file
+ *
+ * @param[in] filename - name of file containing value
+ * @param[out] val - pointer to the result area
+ * @returns True or False depending if value was found
+ *
+ */
+static gboolean 
+readMemoryValueFromFile(const char* filename, guint64* val)
+{
+	gboolean result = FALSE;
+	char *line = NULL;
+	size_t lineLen = 0;
+	char *endptr = NULL;
+	guint64 num = 0, multiplier;
+	FILE *file = NULL;
+
+	if (val != NULL) {
+		file = fopen(filename, "r");
+		if (file != NULL) {
+			if (getline(&line, &lineLen, file) != -1) {
+				errno = 0;
+				num = strtoull(line, &endptr, 0);
+				if (line != endptr && errno == 0) {
+					multiplier = 1;
+
+					switch (*endptr)
```

Comment:
I know this mulitplier handling code is also in coreclr, but it isn't needed. The kernel doesn't return suffixes.

Reply:
Removed